src/Storages/VirtualColumnUtils.cpp

```diff
@@ -214,37 +215,63 @@ static void addPathAndFileToVirtualColumns(Block & block, const String & path, s
         block.getByName("_file").column->assumeMutableRef().insert(file);
     }
 
+    for (const auto & item : partitioning_keys)
+    {
+        if (block.has(item.name))
+        {
+            auto column = block.getByName(item.name).column;
+            ReadBufferFromString buf(item.column->getDataAt(0).toView());
+            item.type->getDefaultSerialization()->deserializeWholeText(column->assumeMutableRef(), buf, getFormatSettings(context));
+        }
+    }
+
     block.getByName("_idx").column->assumeMutableRef().insert(idx);
 }
 
-std::optional<ActionsDAG> createPathAndFileFilterDAG(const ActionsDAG::Node * predicate, const NamesAndTypesList & virtual_columns)
+std::optional<ActionsDAG> createPathAndFileFilterDAG(const ActionsDAG::Node * predicate, const NamesAndTypesList & virtual_columns, const String & path, const ContextPtr & context)
 {
     if (!predicate || virtual_columns.empty())
         return {};
 
     Block block;
+    std::unordered_map<std::string, std::string> keys;
+    if (context->getSettingsRef().use_hive_partitioning)
+        keys = parseHivePartitioningKeysAndValues(path);
     for (const auto & column : virtual_columns)
     {
-        if (column.name == "_file" || column.name == "_path")
+        if (column.name == "_file" || column.name == "_path" || keys.contains(column.name))
```

Comment:
I think we can avoid propogating `path` and using `parseHivePartitioningKeysAndValues` here. We already have all virtual column names in `virtual_columns`. We just should check if this virtual column is known `_path/_file/_size/_time/_etag` or not. If not - treat it as partitioning column

Reply:
We can create function `getCommonVirtualsForFileLikeStorage` that will return `NamesAndTypesList` with columns `_path/_file/_size/_time/_etag`. We can use it in `getVirtualsForFileLikeStorage` and also here